src/archetype.mojo

```
@@ -242,19 +329,84 @@ struct Archetype(CollectionElement, CollectionElementNew):
 
+    fn unsafe_copy_to(self, inout other: Self, index: UInt, other_index: UInt):
+        """Copies all components of the entity at the given index to another archetype.
+
+        Caution: This function does not check whether the other archetype
+                 contains the component.
+
+        Args:
+            other: The archetype to copy the components to.
+            index: The index of the entity.
+            other_index: The index of the entity in the other archetype.
+        """
+        for i in range(self._component_count):
+            other.unsafe_set(
+                other_index,
+                self._ids[i],
+                self._get_component_ptr(index, self._ids[i]),
+            )
+
     @always_inline
-    fn get_component_ptr(
-        self, index: UInt, id: Self.Id
-    ) raises -> UnsafePointer[UInt8]:
-        """Returns the component with the given id at the given index."""
-        self.assert_has_component(id)
+    fn get_component_ptr[
+        assert_has_component: Bool = True
+    ](self, index: UInt, id: Self.Id) raises -> UnsafePointer[UInt8]:
+        """Returns the component with the given id at the given index.
+
+        Args:
+            index:  The index of the entity.
+            id:     The id of the component.
+
+        Parameters:
+            assert_has_component: Whether to assert that the archetype
+                    contains the component.
+
+        Raises:
+            Error:  If assert_has_component and the archetype does not contain the component.
+        """
+
+        @parameter
+        if assert_has_component:
+            self.assert_has_component(id)
         return self._get_component_ptr(index, id)
 
     @always_inline
     fn has_component(self, id: Self.Id) -> Bool:
         """Returns whether the archetype contains the given component id."""
         return bool(self._data[int(id)])
 
+    @always_inline
+    fn has_all_components[T: Intable](self, ids: InlineArray[T]) -> Bool:
```

I don't have a sufficient overview to see where you use this and the following method, but wouldn't the use of masks be more appropriate here?

---

It is used to see if all the components to be removed exist (in `world._remove_and_add`. Using a mask would require us to create and fill a ask first. I agree that a mask would be faster otherwise.

---

For the same reason as detailed in the comment below, I think these methods should not even be required. Only the world is responsible for selecting archetypes, and should not rely on such an expensive check for that sake. End users should not be able to remove components from an inappropriate archtype.

---

The world selects the entity's archetype via the archetype pointer / index stored in the EntityIndex. It therefore does not check the existence of components. If then a component is to be removed from the entity that it does not have, the program should raise rather than having undefined behaviour. Of course, this is not needed in queries or whatever where the archetype is searched by a filter.

---

The different archetype constructors are a bit confusing, but finally you need the archetype's mask to construct it. Creating this mask by altering the mask of the old/original archetype by setting the added/removed bits should be faster than from scratch, and can efficiently cover these checks. And you get the checks for duplicate entries in the add/remove lists for free, which would be relatively costly otherwise. I think you don't have these checks yet.

But go ahead, I don't want to insist further here.

---

I will investigate the case of duplicate components. If not covered already, I will add compile-time checks. 

I am not sure if the thing with altering the original archetype's mask works here, because I do not distinguish between added and removed bits but simply say which bits change. Therefore, I need to assert that the bits were set / unset before. How would you do it otherwise? First do the remove operations and do a bitmask contains with the original mask and then do the addition and do a reverse bitmask contains? That could indeed work. I'll have a look.

---

> I do not distinguish between added and removed bits but simply say which bits change

But isn't that just true for the archetype graph? In `_add_remove`, you know what is added and what is removed.

Mask manipulation and checks are in [World.getExchangeMask](https://github.com/mlange-42/arche/blob/b2a90dd817fa43dc200d5ba5ea99db8a5e48ebb1/ecs/world_internal.go#L533)

The check for duplicate entries is implicit, and therefore for free.

---

OK, I have inserted a compile time check for uniqueness of components. This is free and also checks the program at compile time, which is nicer.

Furthermore, I did the check for adding / removing existent / nonexistent components based on the mask, and I let the getters return (immutable) references to the mask instead of copies (though I still am not sure whether this makes any difference)